From 54341e957a85ef85e70c4c27208c842fea160c68 Mon Sep 17 00:00:00 2001
From: Zev Weiss <zev@bewilderbeest.net>
Date: Thu, 12 Nov 2020 04:16:29 -0600
Subject: [PATCH] Fix edge case with object manager on '/'.

In the special case of the object manager interface being applied on
'/', we need to skip appending the separator before the child path,
since it leads to a double-slash and hence an invalid path, which leads
to panicking down the road.

Fixes: #293
---
 src/stdimpl.rs | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/stdimpl.rs b/src/stdimpl.rs
index eda7c53..5651a69 100644
--- a/src/stdimpl.rs
+++ b/src/stdimpl.rs
@@ -345,7 +345,9 @@ fn get_managed_objects(mut ctx: Context, cr: &mut Crossroads, _: ()) -> Option<C
     let children: Vec<dbus::Path<'static>> =
         cr.get_children(ctx.path()).into_iter().map(|child_path| {
             let mut x = String::from(&**parent);
-            x.push_str("/");
+            if !x.ends_with('/') {
+                x.push_str("/");
+            }
             x.push_str(child_path);
             dbus::Path::from(x).into_static()
         }).collect();
